pilot/pkg/xds/workloadentry.go

```diff
@@ -0,0 +1,233 @@
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"istio.io/api/networking/v1alpha3"
+	"istio.io/istio/pilot/pkg/features"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking/util"
+	"istio.io/istio/pkg/config"
+	"istio.io/istio/pkg/config/schema/gvk"
+)
+
+const (
+	// TODO use status or another proper API instead of annotations
+
+	// AutoRegistrationGroupAnnotation on a WorkloadEntry stores the associated WorkloadGroup.
+	AutoRegistrationGroupAnnotation = "istio.io/autoRegistrationGroup"
+	// WorkloadControllerAnnotation on a WorkloadEntry should store the current/last pilot instance connected to the workload for XDS.
+	WorkloadControllerAnnotation = "istio.io/workloadController"
+	// ConnectedAtAnnotation on a WorkloadEntry stores the time in nanoseconds when the associated workload connected to a Pilot instance.
+	ConnectedAtAnnotation = "istio.io/connectedAt"
+	// DisconnectedAtAnnotation on a WorkloadEntry stores the time in nanoseconds when the associated workload disconnected from a Pilot instance.
+	DisconnectedAtAnnotation = "istio.io/disconnectedAt"
+)
+
+func (sg *InternalGen) RegisterWorkload(proxy *model.Proxy, con *Connection) {
+	// check if the WE already exists, update the status
+	entryName := autoregisteredWorkloadEntryName(proxy)
+	if entryName == "" {
+		return
+	}
+
+	// Try to patch, if it fails then try to create
+	_, err := sg.Store.Patch(gvk.WorkloadEntry, entryName, proxy.Metadata.Namespace, func(cfg config.Config) config.Config {
+		setConnectMeta(&cfg, sg.Server.instanceID, con)
+		return cfg
+	})
+	// TODO return err from Patch through Get
+	if err == nil {
+		return
+	} else if !errors.IsNotFound(err) && err.Error() != "item not found" {
+		adsLog.Warnf("updating auto-registered WorkloadEntry %s/%s: %v", proxy.Metadata.Namespace, entryName, err)
+	}
+
+	// No WorkloadEntry, create one using fields from the associated WorkloadGroup
+	groupCfg := sg.Store.Get(gvk.WorkloadGroup, proxy.Metadata.AutoRegisterGroup, proxy.Metadata.Namespace)
+	if groupCfg == nil {
+		adsLog.Warnf("auto-registration of %v failed: cannot find WorkloadGroup %s/%s", proxy.ID, proxy.Metadata.Namespace, proxy.Metadata.AutoRegisterGroup)
+		return
+	}
+	entry := workloadEntryFromGroup(entryName, proxy, groupCfg)
+	setConnectMeta(entry, sg.Server.instanceID, con)
+	_, err = sg.Store.Create(*entry)
+	if err != nil {
+		adsLog.Errorf("auto-registration of %v failed: error creating WorkloadEntry: %v", proxy.ID, err)
+	}
+}
+
+func (sg *InternalGen) QueueUnregisterWorkload(proxy *model.Proxy) {
+	// check if the WE already exists, update the status
+	entryName := autoregisteredWorkloadEntryName(proxy)
+	if entryName == "" {
+		return
+	}
+	// unset controller, set disconnect time
+	_, err := sg.Store.Patch(gvk.WorkloadEntry, entryName, proxy.Metadata.Namespace, func(cfg config.Config) config.Config {
+		delete(cfg.Annotations, WorkloadControllerAnnotation)
```

Comment:
Is this safe? it may have already reconnected to another instance and had it updated. We need some form of conditional deletion - maybe this is already handled by something via resourceVersion or others though?

Reply:
I suppose we don't need to unset this. The periodic cleanup can just rely on `lastConnect - lastDisconnect`